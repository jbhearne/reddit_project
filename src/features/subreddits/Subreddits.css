
.menu {
    width: 50%;
    border: solid #363376 4px;
    background: white;
    position: absolute;
    left: 25%;
    top: 150px;
    z-index: 10;
    animation-duration: 0.3s;
    animation-name: slidein;
}

@keyframes slidein {
    from {
      margin-top: 100%;
      
    }
  
    to {
      margin-top: 0%;
      
  }
}

.subreddits {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    padding-left: 3%;
}

@media only screen and (max-width: 700px) {
    .subreddits {
        display: grid;
        grid-template-columns: 1fr;
        padding-left: minmax(.5rem, 5rem);
        
    }

    .subreddit div {
        
        
    } 
}

h2 {
    color: aliceblue;
}

.subreddit div {
    display: flex;
    justify-items: left;
    align-items: center;
    
    
}

.subreddit a {
    
}

.subreddit img {
    width: 5vw;
    margin: .25rem 1rem .25rem 15%;
    border: solid #b41e75 3px;
    border-radius: 50% 50% 50% 50%;
}

.subreddit_name {
    font-size: .50rem;
    font-weight: bold;
    color: #8ebd79;
}

.description {
    position: absolute;
    top: 5rem;
    left: 0;
    right: 0;
    background-color: #e2f3f0;
    border: solid #4d8331 4px;
    border-radius: 10px;
    z-index: 11;
    padding: .5rem;
    flex-direction: column;
    
}

.description div {
    flex-direction: column;
    font-size: unset;
}

.description h3 {
    width: 100%;
    
}

div.close {
    font-size: 2rem;
    line-height: .5rem;
    width: 100%;
    text-align: right;
    display: block;
}