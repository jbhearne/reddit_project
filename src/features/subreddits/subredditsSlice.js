import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { computeHeadingLevel } from '@testing-library/react';
import { fetchReddit } from '../../app/redditAPI';

export const fetchSubreddits  = createAsyncThunk(
    'subreddits/fetchSubreddits',
    async (url) => {
        const response = await fetchReddit(url);
        return response.children.map(child => {
            return {
                name: child.data.display_name,
                icon: child.data.icon_img ? child.data.icon_img : '/r.png', //   ./ is different than /  I used ./ and that caused the link to break when the route changed.
                url: child.data.url,
                description: child.data.description,
                id: child.data.id
            }
        })
    }
)

export const subredditsSlice = createSlice({
    name: 'subreddits',
    initialState: {
        url: '', 
        subreddits: [],
        isLoading: false,
        hasError: false
    },
    reducers: {
        setSubredditsUrl: (state, action) => {
            state.url = action.payload
        }
    },
    extraReducers: {
        [fetchSubreddits.pending]: (state, action) => {
            state.isLoading = true;
            state.hasError = false;
        },
        [fetchSubreddits.fulfilled]: (state, action) => {
            state.subreddits = action.payload
            state.isLoading = false;
            state.hasError = false;
        },
        [fetchSubreddits.rejected]: (state, action) => {
            state.isLoading = false;
            state.hasError = true;
        }
    }
})

export const selectSubreddits = (state) => state.subreddits.subreddits;
export const selectSubredditsUrl = (state) => state.subreddits.url;
export const setSubredditsUrl = subredditsSlice.actions.setSubredditsUrl;

export default subredditsSlice.reducer;