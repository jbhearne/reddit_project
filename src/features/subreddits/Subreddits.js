import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { selectSubreddits, selectSubredditsUrl, fetchSubreddits } from './subredditsSlice';
import { Subreddit } from './subreddit/Subreddit';
import './Subreddits.css'

export function Subreddits() {

    const subreddits = useSelector(selectSubreddits);
    const url = useSelector(selectSubredditsUrl);
    const dispatch = useDispatch()

    useEffect(() => {
        dispatch(fetchSubreddits(url))
    }, [url])

    return (
        <div className='menu_container'>
            <div className='menu'>
                <h2>Subreddit</h2> {/*changed to h2 to better align the role*/}
                <ul className='subreddits' role='menu'>
                    {subreddits.map(subreddit => {
                        return (
                            <li className='subreddit' key={subreddit.id}><Subreddit subreddit={subreddit} /></li>
                        )
                    })}
                </ul>
            </div>
        </div>
    )
}