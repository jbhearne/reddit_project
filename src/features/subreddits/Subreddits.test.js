import { useSelector, useDispatch } from "react-redux";
import { selectSubreddits, selectSubredditsUrl, fetchSubreddits } from './subredditsSlice';
import { Subreddits } from './Subreddits'
import { fakeSubredditsComponentData } from '../../../__tests__/fakeComponent.data'
import renderer from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom';
import { NavLink } from 'react-router-dom'

jest.mock('react-redux')
jest.mock('./subredditsSlice')
//jest.mock('react-router-dom')

describe('Subreddits component', () => {
    let mockComponentSubreddits = fakeSubredditsComponentData

    /*fetchSubreddits.mockImplementation(url => {
        mockComponentSubreddits = fakeSubredditsComponentData
    })
    
    useSelector.mockImplementation(selection => {
        if (selection === selectSubreddits) {
            return mockComponentSubreddits
        } else if (selection === selectSubredditsUrl) {
            return 'URLtest'
        } else {
            return 'nope'
        }
    })
    
    useDispatch.mockReturnValue((dis) => dis)*/
    /*it('selectSubreddits', () => {
        useSelector.mockImplementation(selection => {
            if (selection === selectSubreddits) {
                return mockComponentSubreddits
            } else if (selection === selectSubredditsUrl) {
                return 'URLtest'
            } else {
                return 'nope'
            }
        })
        expect(useSelector(selectSubreddits)).toBe(mockComponentSubreddits)
    })*/
    it('should render', () => {
        fetchSubreddits.mockImplementation(url => {
            mockComponentSubreddits = fakeSubredditsComponentData
        })
        
        useSelector.mockImplementation(selection => {
            if (selection === selectSubreddits) {
                return mockComponentSubreddits
            } else if (selection === selectSubredditsUrl) {
                return 'URLtest'
            } else {
                return 'nope'
            }
        })
        
        useDispatch.mockReturnValue((dis) => dis)

        /*NavLink.mockImplementation((props) => { // why can't I mock this? classcomponent
            return (
                <a href={props.to} className={props.className} key={props.key} activeClassName={props.activeClassName}>
                
                </a>
            )
        })*/

        const component = renderer.create(
            <Router>
                <Subreddits />
            </Router>
        )
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot();
        tree = component.toJSON()
        expect(tree).toMatchSnapshot();

    })

})