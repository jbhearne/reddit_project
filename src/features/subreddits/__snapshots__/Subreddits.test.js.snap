// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subreddits component should render 1`] = `
<div
  className="menu_container"
>
  <div
    className="menu"
  >
    <h3>
      Subreddit
    </h3>
    <ul
      className="subreddits"
    >
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUr.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub
          </span>
        </div>
      </li>
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUrl.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon2"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub2
          </span>
        </div>
      </li>
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUrl.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon3"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub3
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Subreddits component should render 2`] = `
<div
  className="menu_container"
>
  <div
    className="menu"
  >
    <h3>
      Subreddit
    </h3>
    <ul
      className="subreddits"
    >
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUr.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub
          </span>
        </div>
      </li>
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUrl.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon2"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub2
          </span>
        </div>
      </li>
      <li
        className="subreddit"
      >
        <div>
          <div
            onClick={[Function]}
          >
            <a
              activeClassName="nav-link-active"
              className="nav-link"
              href="/testSubUrl.json"
              onClick={[Function]}
            >
               
              <img
                className="subreddit_icon"
                src="testSubIcon3"
              />
            </a>
          </div>
          <span
            className="subreddit_name"
            onClick={[Function]}
          >
            testSub3
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
